lib/msal-core/src/UserAgentApplication.ts

```diff
@@ -1111,18 +1111,22 @@ export class UserAgentApplication {
             // Redirect to home page if login request url is null (real null or the string null)
             if (!loginRequestUrl || loginRequestUrl === "null") {
                 this.logger.error("Unable to get valid login request url from cache, redirecting to home page");
-                window.location.href = "/";
+                window.location.assign("/");
                 return;
-            } else if (currentUrl !== loginRequestUrl) {
-                // If loginRequestUrl contains a hash (e.g. Angular routing), process the hash now then redirect to prevent both hashes in url
-                if (loginRequestUrl.indexOf("#") > -1) {
-                    this.logger.info("loginRequestUrl contains hash, processing response hash immediately then redirecting");
-                    this.processCallBack(hash, stateInfo, null);
-                    window.location.href = loginRequestUrl;
+            } else {
+                if (!UrlUtils.isSamePage(currentUrl, loginRequestUrl)) {
+                    const finalRedirectUrl = UrlUtils.getBaseUrl(loginRequestUrl);
+                    window.location.assign(`${finalRedirectUrl}${hash}`);
+                    return;
                 } else {
-                    window.location.href = `${loginRequestUrl}${hash}`;
+                    const loginRequestUrlComponents = UrlUtils.GetUrlComponents(loginRequestUrl);
+                    if (loginRequestUrlComponents.Hash){
+                        window.location.hash = loginRequestUrlComponents.Hash;
+                    }
+                    if (loginRequestUrlComponents.Search){
+                        window.location.search = loginRequestUrlComponents.Search;
```

Comment:
Changing `window.location.search` actually initiates a browser navigation, while changing `window.location.hash` does not. I think you may want to put `processCallback` before you change `search`, to prevent a race condition?

Reply:
In other words, the order in which you call process hash versus changing the `window.location.hash` does not matter, but it technically matters if you are changing `window.location.search`, since it could navigate away from the page completely. I think there should be separate behavior depending on if there is a query parameter, instead of relying on `this.processCallback` at the end of the function.

Reply:
Oh interesting. I didn't experience another navigation in my testing. Is this only in some scenarios? I could just move the url manipulation block underneath the `processCallback` call. That makes sense to me.